### ISSUE TITLE
Port pysal to python3

### ISSUE THREAD

**Issue by sjsrey** [pysal/pysal#151]
_Original author: phil.stp...@gmail.com (April 13, 2011 22:22:52)_

<b>Task to be completed</b>

http://pysal.org/1.2/developers/pep/pep-0009.html

_Original issue: http://code.google.com/p/pysal/issues/detail?id=151_

**Comment by giumas** [comment]
Do you have a schedule for this task?

**Comment by sjsrey** [comment]
@giumas We would certainly like to have it done in the next release or two. It is a function of having many demands on a finite set of devs, so any additional help would move this forward.

**Comment by giumas** [comment]
@sjsrey  I have forked to see how big is the issue. Is '3convert' the correct  branch for working on this topic?

**Comment by sjsrey** [comment]
Great to hear.

We have notes [here](https://github.com/pysal/pysal/wiki/Converting-2-to-3-Notes) on how we have been working on the 2-3 port. 

The 3convert branch is a bit stale so it might be best just to work off a new branch from master. There will be others likely who want to work on the port this release cycle so they may add in some ideas as well.

**Comment by giumas** [comment]
I read these notes.. Do you really need all the modernize+2to3 steps? Following these simple directions I am getting good results: http://python-future.org/compatible_idioms.html 
What is the minimum python 2.x that you want to maintain?

**Comment by ljwolf** [comment]
Hey, @giumas, thanks for the help! Maybe this will help clear things up.

Python-modernize uses `six` and `future` in a way that swaps things into 2.7 idioms that are compatible with python3. This is so that the amount of code that changes in the `2to3` conversion is minimized. This approach is replaces 2-only idioms with ones that are compatible with both 2 and 3, like the site you've suggested recommends.  Then, we use `2to3` and test the python3 code.

Altogether, this is just another process that does what the `futurize` function on `python-future` does. If `futurize` gives better results, we can swap to using that for automating getting the code modernized. It's the same process, just a different tool.

That said, I've tested your fork using `nosetests-3.4` and some of the [errors] (https://gist.github.com/ljwolf/1c64fdaad2b38c8e1d7c#file-gistfile1-txt-L984) in this log point to a few of the same core issues we've been working on to get PySAL on Python 3.

FileIO is the tricky wicket. Once we fix for the metaclassing issue that arises in the above `nosetest` output (`AttributeError: type object 'FileIO' has no attribute 'FileIO'`), there are still problems.

 Specifically, we use a lot of `StringIO` and `struct` to pack and unpack bytestrings. This  needs to get corrected by hand. The current implementation's reliance on properties of 2.*'s StringIO interfaces don't convert well because automated tools have no string introspection capability.

Regrettably, the hard lifting seems to me to be in fixing core issues in the IO handling of dbf and shapefiles, because, as you've identified, most of the rest of the library appears to be python3-ready.

If you just grep through that nosetest log looking for the FileIO errors, you'll see that most of the issues are related to this one module, its metaclass, and its use of `struct` and `StringIO`. If you can get your data *in*, then pysal should be able to operate on it in a data-agnostic way.

**Comment by ljwolf** [comment]
whoops, didn't mean to close!

**Comment by giumas** [comment]
My fork was exploratory.. just to understand a bit the job size. A fresh branch with the preliminary 'mechanical' steps done might be useful.

**Comment by ljwolf** [comment]
Exploring this is really helpful. Fresh eyes are always good! We'll have a new `3convert` up soon, and hacking on that branch might yield more fruit.

**Comment by marwahaha** [comment]
What's the status on porting to Python3? Is there any way I can help?

**Comment by ljwolf** [comment]
As it stands, all CI tests are passing on @sjsrey's staging branch, [py3merge] (https://github.com/sjsrey/pysal/tree/py3merge), and we are looking at merging it soon. 

It would be super helpful if you could try to install it from that branch with `pip3`:
`pip3 install git+https://github.com/sjsrey/pysal.git@py3merge`
or by downloading that branch and executing its `setup.py` with a python `3.*` version. 

Some have done this before, and no one has reported errors. But, more is better here!

**Comment by marwahaha** [comment]
I've installed it, with no errors! (Windows 8 + Git Bash). Anything you'd like me to try?

**Comment by sjsrey** [comment]
Thanks @marwahaha for giving this a test. I saw that you updated a notebook link https://github.com/pysal/notebooks/commit/9572972043e88aca95e89dde07da001a87505f95 - if you could run through the notebooks using the py3merge install and highlight any errors that may crop up, that would be very helpful (if you haven't already).

**Comment by sjsrey** [comment]
I tested the py3merge on the notebooks from the NARSC2015 workshop. There were a bunch of `s/xrange/range/`   as well as `s/print thing/print(thing)/` changes. The only "substantive" change had to do with [iterating over shapes in our reader](https://github.com/sjsrey/pysal_narsc_2015/commit/6e237e8ba6652f49abf58862d09fc1934861eaf8#diff-f0c1ae1446b2eb016e979d2786f0a744L736). It was balking about the reader not having a `next`method. So I changed the call to use `.get(i)`.

I checked our weights builders and they are all fine so this was an issue with one explicit approach to stepping over the shapes (which we don't use internally).

**Comment by ljwolf** [comment]
Actually, instead of `s/xrange/range`, I'd suggest doing this at top:
```
if sys.version_info[0] > 2:
    xrange = range
```
That way, in python2, you're still using the efficient `xrange`, and in python3 you're using the default `range` (which is xrange).

**Comment by jlaura** [comment]
+1 really good call - good to know!

**Comment by slaughterhouse-give** [comment]
pip3 install git+https://github.com/sjsrey/pysal.git@py3merge worked for me. Python3.3.0, CentOS release 6.7

**Comment by sjsrey** [comment]
If anybody wants to test this on a python 3 environment, there is a test build that can be used:

`pip install -i https://testpypi.python.org/pypi pysal`

**Comment by marwahaha** [comment]
Hi, I started going through the notebooks, but many had older versions of IPython, so I think most of the errors are due to the notebooks being old. I do have some places where I don't understand the errors --- shall I post it here or make a new issue?

**Comment by pastephens** [comment]
@Levi have you run across a way to include the notebooks in a test suite?

**Comment by sjsrey** [comment]
@marwahaha thanks for testing this. I also saw some errors in the notebooks, mostly having to do with range vs. xrange (see earlier in this thread). I think if you post them here, we can work on the 3 port and then if they turn out to be due to the notebooks being out of date, we can then fix them.

**Comment by ljwolf** [comment]
@pastephens nope, I haven't. A quick look shows [this gist](https://gist.github.com/minrk/2620735). Totally forgot that we'd need to convert the notebooks as well, and I don't know of an automated tool to move them into the compatible 2/3 subset...

**Comment by marwahaha** [comment]
I ran the pysal weights notebook in Python 2, and still received one error:
https://github.com/pysal/notebooks/blob/master/notebooks/PySAL_weights.ipynb
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-112-27ab1032f004> in <module>()
----> 1 wknn3 = ps.knnW(data, k = 3)

C:\Anaconda\lib\site-packages\pysal\weights\Distance.pyc in knnW(kdtree, k, p, ids)
     84     """
     85     data = kdtree.data
---> 86     nnq = kdtree.query(data, k=k+1, p=p)
     87     info = nnq[1]
     88 

AttributeError: 'numpy.ndarray' object has no attribute 'query'
```

**Comment by sjsrey** [comment]
That is due to a change in the way distance weights are handled, and not due to the change to support Python 3.

`ps.knnW` now takes a KDTree not an array. I will update the notebook shortly.

For now cell 74 in the notebook should read:

```
tree = ps.cg.kdtree.KDTree(np.array(data))
wknn3 = ps.knnW(tree, k = 3)
```

**Comment by marwahaha** [comment]
The link is broken in intro_scicomp (are we maintaining this notebook?)
```
data_link = 'http://ubuntuone.com/4oIpVJDCpdREhzNdSRwMx8'
db = pd.read_csv(data_link)
db
```

**Comment by marwahaha** [comment]
On the last line of PySAL_esda, I can't get this working in Python 2
```
hoods = viz.Map_Projection("data/amsterdam_hoods.shp")
viz.lisa_cluster_map(hoods, lisa_total, title="LISA Cluster Map")
```

**Comment by marwahaha** [comment]
^I fixed the above to be 
```
hoods = "data/amsterdam_hoods.shp"
viz.plot_lisa_cluster(hoods, lisa_total, title="LISA Cluster Map")
```
but in Python 3, this fails. I think it's the ```setup_ax``` call.
```
C:\Anaconda\envs\py3\lib\site-packages\matplotlib\artist.py:210: MatplotlibDeprecationWarning: This has been deprecated in mpl 1.5, please use the
axes property.  A removal date has not been set.
  warnings.warn(_get_axes_msg, mplDeprecation, stacklevel=1)
C:\Anaconda\envs\py3\lib\site-packages\matplotlib\artist.py:221: MatplotlibDeprecationWarning: This has been deprecated in mpl 1.5, please use the
axes property.  A removal date has not been set.
  warnings.warn(_get_axes_msg, mplDeprecation, stacklevel=1)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-26-3536e8dab3a4> in <module>()
      1 hoods = "data/amsterdam_hoods.shp"
----> 2 viz.plot_lisa_cluster(hoods, lisa_total, title="LISA Cluster Map")

C:\Anaconda\envs\py3\lib\site-packages\pysal\contrib\viz\mapping.py in plot_lisa_cluster(shp_link, lisa, p_thres, shp_type, title, legend, savein, figsize, dpi, alpha, leg_loc)
    711     fig = plt.figure(figsize=figsize)
    712     ax = fig.add_subplot(111)
--> 713     ax = setup_ax([lisa_obj], ax)
    714     # Legend
    715     if legend:

C:\Anaconda\envs\py3\lib\site-packages\pysal\contrib\viz\mapping.py in setup_ax(polyCos_list, ax)
    223         polyCo.get_axes().set_xlim(ax.get_xlim())
    224         polyCo.get_axes().set_ylim(ax.get_ylim())
--> 225         ax.add_collection(polyCo)
    226     ax.set_frame_on(False)
    227     ax.axes.get_yaxis().set_visible(False)

C:\Anaconda\envs\py3\lib\site-packages\matplotlib\axes\_base.py in add_collection(self, collection, autolim)
   1664             collection.set_label('_collection%d' % len(self.collections))
   1665         self.collections.append(collection)
-> 1666         self._set_artist_props(collection)
   1667 
   1668         if collection.get_clip_path() is None:

C:\Anaconda\envs\py3\lib\site-packages\matplotlib\axes\_base.py in _set_artist_props(self, a)
    863             a.set_transform(self.transData)
    864 
--> 865         a.axes = self
    866         if a.mouseover:
    867             self.mouseover_set.add(a)

C:\Anaconda\envs\py3\lib\site-packages\matplotlib\artist.py in axes(self, new_axes)
    235         if (new_axes is not None and
    236                 (self._axes is not None and new_axes != self._axes)):
--> 237             raise ValueError("Can not reset the axes.  You are "
    238                              "probably trying to re-use an artist "
    239                              "in more than one Axes which is not "

ValueError: Can not reset the axes.  You are probably trying to re-use an artist in more than one Axes which is not supported
```

**Comment by sjsrey** [comment]
The `ValueError: Can not reset the axes.` error has been fixed, but after I built the 3.0 source distribution. I will rebuild it (likely tomorrow) and that should handle this error.

**Comment by darribas** [comment]
@marwahaha On the broken link to the file on Ubuntu One: that was a file of tweets used for a specific workshop but their license do not allow to redistribute, so I can't upload it to Github to link to it I'm afraid. At least the notebook should serve as illustration.

**Comment by sjsrey** [comment]
@marwahaha and @darribas i updated the testing 3 version of pysal on testing pypi which can be installed with:

`pip install -i https://testpypi.python.org/pypi pysal`

running this on the updated notebooks from https://github.com/pysal/notebooks/pull/22 looks good (other than the twitter data permission thing).

haven't tested on python 2 yet, though.

**Comment by sjsrey** [comment]
Tested the source candidate above on Python 2.7 via pip and all worked.

**Comment by marwahaha** [comment]
Great. What else would you like me to test? I'm very interested in PySAL for Python 3 :-)

**Comment by sjsrey** [comment]
@marwahaha Thanks for the help. We are waiting on a few more additions from the team that will get rolled into the release which is planned for 1/30. At that point there will be official support for Python 3 :->.

**Comment by ljwolf** [comment]
Release 1.11 now supports python 3.4 and above.
